accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,706 @@
+bool f = arrInt is delegate*<string>[]; // 'false'
+```
+which uses `isinst`:
+```
+.locals init (object V_0, bool V_1)
+ldc.i4.1
+newarr     method int32 *()
+stloc.0
+ldloc.0
+isinst     method string *()[]
+ldnull
+cgt.un
+stloc.1
+```
+
+### `Is` keyword with known values
+C# will not use `isinst` when it "knows" the types are equal or not equal:
+```cs
+var fn = new delegate*<int>[1];
+Console.WriteLine(fn is delegate*<int>[]); // true
+Console.WriteLine(fn is delegate*<bool>[]); // false
+```
+```
+ldc.i4.1
+newarr     method int32 *()
+ldnull
+cgt.un
+call       void [System.Console]System.Console::WriteLine(bool)
+ldc.i4.0
+call       void [System.Console]System.Console::WriteLine(bool)
+```
+
+The C# compiler will also consider the `false` case warning CS0184: `The given expression is never of the provided...`.
+
+This C# vs. runtime disagreement with "known values" also appears with array variance:
+```cs
+Console.WriteLine(new int[1] is uint[]); // false
+Console.WriteLine(((object)new int[1]) is uint[]); // true
+```
+
+### With modifiers + unmodified types
+The above examples only compare parameter types without calling conventions or modifiers.
+
+An unmodified type does include managed vs. unmanaged as part of its type identity:
+```cs
+object o = new delegate*<int>[1];
+Console.WriteLine(o is delegate*<int>[]); // true
+Console.WriteLine(o is delegate* unmanaged[Cdecl]<int>[]); // false
+
+var fn = new delegate*<int>[1];
+Console.WriteLine(fn is delegate*<int>[]); // true
+Console.WriteLine(fn is delegate* unmanaged[Cdecl] <int>[]); // false
+```
+
+but unmanaged calling conventions are not part of the type identity:
+```cs
+object o1 = new delegate* unmanaged[SuppressGCTransition]<int>[1];
+Console.WriteLine(o1 is delegate* unmanaged[MemberFunction]<int>[]); // true
+
+object o2 = new delegate* unmanaged[CDecl]<int>[1];
+Console.WriteLine(o2 is delegate* unmanaged[StdCall]<int>[]); // true
+```
+
+In addition, all non-CallConv* modifiers, such as `ref\in\out\const` are not considered part of the type identity either.
+
+### With modified types
+A modified type acts like the underlying type with the exception that it overrides `GetFunctionPointerParameters()` to return the modified parameter types. It has an identity separate than the underlying types.
+
+```cs
+public class Holder
+{
+    public unsafe delegate* unmanaged[SuppressGCTransition]<int> _field;
+}
+
+Type modifiedType = typeof(Holder).GetField("_field").GetModifiedFieldType();
+Console.WriteLine(modifiedType.IsModifiedType); // true
+Console.WriteLine(modifiedType.UnderlyingSystemType); // (unmodified function pointer)
+Console.WriteLine(modifiedType.IsFunctionPointer); // false
+Console.WriteLine(modifiedType.UnderlyingSystemType.IsFunctionPointer); // true
+
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions().Length); // 2
+Console.WriteLine(modifiedType.GetFunctionPointerCallingConventions[0]); // CallConvCdecl
```

Comment:
Where did `CallConvCdecl` came from? The function pointer above uses default calling convention, it does not use Cdecl.

Reply:
Oops, typos caused by refactoring that section earlier. Updated.